crates/neon-sys/native/src/neon.cc

```diff
@@ -541,3 +541,19 @@ extern "C" void Neon_EventHandler_Delete(void * thread_safe_cb) {
     neon::EventHandler *cb = static_cast<neon::EventHandler*>(thread_safe_cb);
     cb->close();
 }
+
+extern "C" void* Neon_TryCatch_New() {
+  return new Nan::TryCatch();
```

Comment:
Unfortunately this isn't going to work. :( a `TryCatch` [has to be allocated on the stack](https://v8docs.nodesource.com/node-8.16/d4/dc6/classv8_1_1_try_catch.html#af28a8ba7e1cb6e66bc2468faa4c86eec). There are a couple ways to do this, both of which are pretty annoying:
- Use a cross-language callback pattern, where you pass the Rust closure as a `void*` to C along with a Rust glue function that takes the Rust closure as an argument and calls in. Then in C you stack-allocate the `TryCatch` and invoke the Rust glue function.
- Allocate a `TryCatch` on the Rust stack by creating a parallel Rust struct type that's big enough to contain the sizeof the C++ `TryCatch`. For this you have to have some way at build time to determine the sizeof a `TryCatch` and use a macro or a const declaration to declare the Rust type. And then the constructor and destructor for that Rust type can call into `neon_runtime` with a pointer to `self` for the C++ code to explicitly call the C++ constructor and destructor on it. This is how `HandleScope` works, which also has to be stack allocated.

Reply:
I believe that the reason they should be stack allocated is for nesting to work properly they need to be deallocated in the same order they were created.

RII also enforces this. I think deleting with `Drop` would also accomplish this.

With that said, I think your suggestion of passing the Rust closure to be invoked by C is best.

Reply:
Hmmm, panicking across FFI is undefined.

Reply:
Oh, this is neat. https://github.com/rust-lang/rust/pull/65646